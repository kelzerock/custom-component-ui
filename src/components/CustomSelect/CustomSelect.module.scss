@mixin shrinkSpan {
  top: 0px;
  left: 5px;
  padding: 1px 3px;
  font-size: .8rem;
  border-radius: 3px;
  transform: translateY(-50%);
  background-color: var(--color-white);
}

.defaultOption {
  display: none;
}

.wrapper {
  position: relative;
  margin: 10px;
}

.title {
  position: absolute;
  left: 10px;
  top: 50%;
  transform: translateY(-50%);
  transition: top 300ms ease, font-size 300ms ease;
}

.select {
  min-height: 30px;
  min-width: 120px;
  border-radius: 5px;

  &:focus {
    border-color: var(--color-main-blue);
    outline: none;
  }
}

.select:focus~.title {
  @include shrinkSpan();
  color: var(--color-main-blue);
}

.select.shrink~.title {
  @include shrinkSpan();
}

.option {}