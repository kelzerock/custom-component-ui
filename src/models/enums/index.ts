export { Checkbox } from "./enums-for-test/Checkbox";
export { Switch } from "./enums-for-test/Switch";
export { Modal } from "./enums-for-test/Modal";
