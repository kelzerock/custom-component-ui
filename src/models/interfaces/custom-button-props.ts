import { ButtonHTMLAttributes, ReactNode } from "react";
import { ButtonSize, ButtonVariant } from "../types";

export interface CustomButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement> {
  children?: ReactNode;
  variant?: ButtonVariant;
  size?: ButtonSize;
}
