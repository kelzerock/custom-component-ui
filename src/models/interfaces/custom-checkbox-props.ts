import { InputHTMLAttributes } from "react";

export interface CustomCheckboxProps
  extends InputHTMLAttributes<HTMLInputElement> {
  title?: string;
  color?: string;
}
