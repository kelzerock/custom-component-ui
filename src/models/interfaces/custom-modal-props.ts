import { HTMLAttributes, ReactNode } from "react";

export interface CustomModalProps extends HTMLAttributes<HTMLDivElement> {
  children?: ReactNode;
  open: boolean;
  onClose: () => void;
}
