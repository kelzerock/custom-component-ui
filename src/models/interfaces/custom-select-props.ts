import { SelectHTMLAttributes } from "react";
import { OptionsData } from "./option-data";

export interface CustomSelectProps
  extends SelectHTMLAttributes<HTMLSelectElement> {
  title?: string;
  options?: OptionsData[];
}
