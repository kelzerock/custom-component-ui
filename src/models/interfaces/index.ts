export { CustomSelectProps } from "./custom-select-props";
export { CustomButtonProps } from "./custom-button-props";
export { CustomTextFieldProps } from "./custom-text-field-props";
export { OptionsData } from "./option-data";
