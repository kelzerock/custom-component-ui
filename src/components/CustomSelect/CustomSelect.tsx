import { ChangeEvent, FC, useMemo, useState } from "react";
import { CustomSelectProps, OptionsData } from "../../models/interfaces";
import styles from "./CustomSelect.module.scss";
import clsx from "clsx";

export const CustomSelect: FC<CustomSelectProps> = ({
  options = [],
  title,
  ...props
}: CustomSelectProps) => {
  const defaultSelected = useMemo(
    () => options.find((el) => el.defaultSelect === true) || null,
    [options],
  );
  const [selectedValue, setSelectedValue] = useState<OptionsData | null>(
    defaultSelected,
  );
  const handleChange = (event: ChangeEvent<HTMLSelectElement>) => {
    console.log("start");
    if (props.disabled) return;

    const selectedOption = options.find((opt) => {
      console.log(opt.value, event.target.value);

      return String(opt.value) === event.target.value;
    });
    console.log({ selectedOption });
    setSelectedValue(selectedOption || null);

    event.target.classList.toggle(styles.shrink, event.target.value !== "");
  };

  return (
    <div
      className={clsx(styles.wrapper, props.disabled && styles.disabled)}
      data-testid="wrapper"
    >
      <select
        className={clsx(styles.select, defaultSelected && styles.shrink)}
        value={selectedValue?.value || ""}
        onChange={handleChange}
        data-testid="select"
      >
        {defaultSelected === null && (
          <option
            value=""
            className={styles.defaultOption}
            disabled
            data-testid="defaultOption"
          ></option>
        )}
        {options.map((option) => (
          <option key={option.value} value={option.value} data-testid="option">
            {option.title}
          </option>
        ))}
      </select>
      {title && (
        <span className={styles.title} data-testid="title">
          {title}
        </span>
      )}
    </div>
  );
};
