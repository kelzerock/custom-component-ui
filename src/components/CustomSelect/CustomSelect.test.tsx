import { render, screen } from "@testing-library/react";
import { CustomSelect } from "./CustomSelect";
import { OptionsData } from "../../models/interfaces";

describe("check CustomSelect", () => {
  test("check title correctly rendered", () => {
    const value = "test value";
    render(<CustomSelect title={value} />);
    const title = screen.getByTestId("title");
    expect(title).toBeInTheDocument();
    expect(title).toHaveTextContent(value);
    expect(title).toHaveClass("title");
  });

  test("check title doesn't render without props", () => {
    render(<CustomSelect />);
    const title = screen.queryByTestId("title");
    expect(title).toBeNull();
  });

  test("check options correct render", () => {
    const options: OptionsData[] = [
      {
        id: 1,
        title: "one",
        value: 1,
      },
      {
        id: 2,
        title: "two",
        value: 2,
      },
      {
        id: 3,
        title: "three",
        value: 3,
      },
    ];
    render(<CustomSelect options={options} />);
    const optionsTest =
      (screen.queryAllByTestId("option") as HTMLOptionElement[]) || null;
    expect(
      optionsTest.every((option) =>
        options.some(
          (el) =>
            String(el.title) === option.textContent &&
            String(el.value) === option.value,
        ),
      ),
    ).toBe(true);
    const defaultOption = screen.getByTestId(
      "defaultOption",
    ) as HTMLOptionElement;
    expect(
      defaultOption.value === "" && defaultOption.textContent === "",
    ).toBeTruthy();
  });

  test("check UI snapshot", () => {
    const { container } = render(<CustomSelect title="age" />);
    expect(container).toMatchSnapshot();
  });
});
