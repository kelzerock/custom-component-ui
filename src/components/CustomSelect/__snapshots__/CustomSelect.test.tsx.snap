// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check CustomSelect check UI snapshot 1`] = `
<div>
  <div
    class="wrapper"
    data-testid="wrapper"
  >
    <select
      class="select"
      data-testid="select"
    >
      <option
        class="defaultOption"
        data-testid="defaultOption"
        disabled=""
        value=""
      />
    </select>
    <span
      class="title"
      data-testid="title"
    >
      age
    </span>
  </div>
</div>
`;
